### ISSUE TITLE
The results I reproduced are far worse than those reported in the paper.

### ISSUE THREAD

**Issue by Tuesdayo** [Michel-liu/FatFormer#8]
This may be due to the data preprocessing methods or the training strategies. Could you share the code for loading the training data and the training process?

**Comment by Cwyxx** [comment]
You need to properly configure the inference parameters, specifically num_vit_adapter=3 and num_context_embedding=8. It works for me.

**Comment by Tuesdayo** [comment]
Can you share your training code? All my configurations are the same as those provided by the author. But the reproduction results were very poor.



---Original---
From: "Chen ***@***.***&gt;
Date: Sun, Mar 30, 2025 11:14 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Michel-liu/FatFormer] The results I reproduced are far worsethan those reported in the paper. (Issue #8)


   
You need to properly configure the inference parameters, specifically num_vit_adapter=3 and num_context_embedding=8. It works for me.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;
  Cwyxx left a comment (Michel-liu/FatFormer#8)
 
You need to properly configure the inference parameters, specifically num_vit_adapter=3 and num_context_embedding=8. It works for me.

**Comment by Cwyxx** [comment]
I’m simply using the checkpoint that the author has made available.

**Comment by Cwyxx** [comment]
biggan accuracy is 99%.

**Comment by Tuesdayo** [comment]
I trained using training code that I wrote myself or from other papers from the same team as the author. The results are very bad. I sent the author a request for training code but got no response.



---Original---
From: "Chen ***@***.***&gt;
Date: Sun, Mar 30, 2025 11:23 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [Michel-liu/FatFormer] The results I reproduced are far worsethan those reported in the paper. (Issue #8)


   
I’m simply using the checkpoint that the author has made available.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;
  Cwyxx left a comment (Michel-liu/FatFormer#8)
 
I’m simply using the checkpoint that the author has made available.

**Comment by Cwyxx** [comment]
sad!😂

**Comment by jungkoak** [comment]
> I trained using training code that I wrote myself or from other papers from the same team as the author. The results are very bad. I sent the author a request for training code but got no response.
> […](#)
> 
> ---Original---
> From: "Chen ***@***.***&gt;
> Date: Sun, Mar 30, 2025 11:23 AM
> To: ***@***.***&gt;;
> Cc: ***@***.******@***.***&gt;;
> Subject: Re: [Michel-liu/FatFormer] The results I reproduced are far worsethan those reported in the paper. (Issue [#8](https://github.com/Michel-liu/FatFormer/issues/8))
> 
> 
> 
> I’m simply using the checkpoint that the author has made available.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***&gt;
>   Cwyxx left a comment ([Michel-liu/FatFormer#8](https://github.com/Michel-liu/FatFormer/issues/8))
> 
> I’m simply using the checkpoint that the author has made available.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

I also sent an email to the author requesting the training code, and there was no message...

**Comment by LiJie77777** [comment]
@Tuesdayo @Cwyxx Pls can you share the FAA module code thanks

**Comment by Cwyxx** [comment]
just follow the main.py

**Comment by Cwyxx** [comment]
`import os
import torch
import random
import numpy as np
import argparse
from utils.misc import get_rank, init_distributed_mode
import utils.misc as utils
from models import build_model
import torchvision.transforms as transforms
from torch.utils.data import Dataset, DataLoader
from sklearn.metrics import average_precision_score, accuracy_score
from PIL import Image
from tqdm import tqdm
import pandas as pd


def recursively_read(
    rootdir, exts=["png", "jpg", "jpeg", "bmp", "webp"]
):
    out = []
    if rootdir is None:
        return out
    for r, d, f in os.walk(rootdir):
        for file in f:
            if (file.split(".")[1].lower() in exts):
                out.append(os.path.join(r, file))
    return out


def get_list(path):
    if path is None:
        return []
    image_list = recursively_read(path)
    return image_list

class RealFakeDataset(Dataset):
    def __init__(
        self,
        real_path,
        fake_path,
        transform,
        get_image_path=False,
    ):
        real_list, fake_list = self.read_path(real_path, fake_path)
        self.total_list = real_list + fake_list

        self.labels_dict = {}
        for i in real_list:
            self.labels_dict[i] = 0
        for i in fake_list:
            self.labels_dict[i] = 1
        self.transform = transform
        self.get_image_path = get_image_path

    def read_path(self, real_path, fake_path):
        real_list = get_list(real_path)
        fake_list = get_list(fake_path)
        print("real: %d, fake: %d" % (len(real_list), len(fake_list)))
        return real_list, fake_list

    def __len__(self):
        return len(self.total_list)

    def __getitem__(self, idx):
        img_path = self.total_list[idx]
        label = self.labels_dict[img_path]
        img = Image.open(img_path).convert("RGB")
        img = self.transform(img)
        if self.get_image_path:
            return img, label, img_path
        return img, label

def get_args_parser():
    parser = argparse.ArgumentParser('Set transformer detector', add_help=False)
        
    parser.add_argument('--seed', type=int, default=100)
    parser.add_argument('--device', default='cuda',
                        help='device to use for training / testing')

    # dataset parameters
    parser.add_argument('--data_dir', type=str, default=None)
    parser.add_argument('--data_split', type=str, default=None)
    parser.add_argument('--generator_list', type=str, default=None)
    parser.add_argument("--result_folder", type=str, default=None)
    parser.add_argument('--img_resolution', type=int, default=256)
    parser.add_argument('--crop_resolution', type=int, default=224)
    parser.add_argument('--batchsize', type=int, default=32)

    # model
    parser.add_argument('--backbone', type=str, default='CLIP:ViT-L/14')
    parser.add_argument('--num_classes', type=int, default=2)
    parser.add_argument('--num_vit_adapter', type=int, default=8)

    # text encoder
    parser.add_argument('--num_context_embedding', type=int, default=8)
    parser.add_argument('--init_context_embedding', type=str, default="")

    # frequency
    parser.add_argument('--hidden_dim', type=int, default=768)
    parser.add_argument('--clip_vision_width', type=int, default=1024)
    parser.add_argument('--frequency_encoder_layer', type=int, default=2)
    parser.add_argument('--decoder_layer', type=int, default=4)
    parser.add_argument('--num_heads', type=int, default=12)

    # output
    parser.add_argument('--eval', action='store_true')
    parser.add_argument('--pretrained_model', type=str, default="/data3/chenweiyan/2024-12/2025-1/code/github/FatFormer/pretrained/fatformer_4class_ckpt.pth")
    parser.add_argument('--num_workers', default=0, type=int)
    parser.add_argument('--print_freq', default=50, type=int)

    # distributed training parameters
    parser.add_argument('--world_size', default=1, type=int,
                        help='number of distributed processes')
    parser.add_argument('--dist_url', default='env://', help='url used to set up distributed training')
    parser.add_argument('--rank', default=0, type=int,
                        help='number of distributed processes')
    parser.add_argument("--local_rank", type=int, help='local rank for DistributedDataParallel')

    return parser


@torch.no_grad()
def evaluate(model, data_loader, device, args=None):
    y_true, y_pred, all_uids = [], [], []
    for imgs, labels, image_paths in tqdm(data_loader):
        imgs = imgs.to(device)
        outputs = model(imgs)
        
        y_pred.extend(outputs.softmax(dim=1)[:, 1].flatten().tolist())
        y_true.extend(labels.flatten().tolist())
        
        uids = [os.path.splitext(os.path.basename(image_path))[0] for image_path in image_paths]
        all_uids.extend(uids)

    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    r_acc = accuracy_score(y_true[y_true==0], y_pred[y_true==0] > 0.5)
    f_acc = accuracy_score(y_true[y_true==1], y_pred[y_true==1] > 0.5)
    acc = accuracy_score(y_true, y_pred > 0.5)
        
    return r_acc, f_acc, acc, y_pred, y_true, all_uids


def main(args):
    fatformer_transform = transforms.Compose([
                transforms.Resize((args.img_resolution, args.img_resolution)),
                transforms.CenterCrop(args.crop_resolution),
                transforms.ToTensor(),
                transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
            ])

    # set device
    init_distributed_mode(args)
    device = torch.device(args.device)

    # fix the seed
    seed = args.seed
    torch.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.cuda.manual_seed(seed)
    
    # model
    model = build_model(args)
    model = model.to(device)
    model_without_ddp = model
    if args.distributed:
        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.gpu], find_unused_parameters=False)
        model_without_ddp = model.module
    
    if args.eval:
        checkpoint = torch.load(args.pretrained_model, map_location='cpu')
        model_without_ddp.load_state_dict(checkpoint['model'], strict=False)
        
    dfs = []
    generator_list = [ generator.strip() for generator in args.generator_list.strip().split(',')]
    
    for generator in generator_list:
        data_dir = os.path.join(args.data_dir, generator)
        if generator == "real" or "real" in generator:
            real_path = data_dir
            fake_path = None
        else:
            real_path = None
            fake_path = data_dir
        
        print(f"real_path: {real_path}")
        print(f"fake_path: {fake_path}")
        dataset = RealFakeDataset(
            real_path=real_path,
            fake_path=fake_path,
            transform=fatformer_transform,
            get_image_path=True,
        )
        dataloader = DataLoader(dataset, batch_size=args.batchsize, shuffle=False)
        r_acc, f_acc, acc, preds, labels, uids = evaluate(model, dataloader, device, args=args)
        
        print(f"{generator}")
        print(f"\tr_acc:{r_acc}")
        print(f"\tf_acc:{f_acc}")
        print(f"\ttotal_acc:{acc}")
        
        df = pd.DataFrame({"generator": generator, "uid": uids, "pred": preds, "label": labels})
        dfs.append(df)
        
    df = pd.concat(dfs, ignore_index=True)
    os.makedirs(args.result_folder, exist_ok=True)
    csv_dir = os.path.join(args.result_folder, f"{args.data_split}.csv")
    df.to_csv(csv_dir, index=False)


if __name__ == "__main__":    
    parser = get_args_parser()
    args = parser.parse_args()
    main(args)
`

**Comment by LiJie77777** [comment]
@Cwyxx Thanks, Forgive me for being a newbie, I didn't see the code referenced by dwt, please give me some guidance again.
#I see that this paragraph is not used
main.py frequency
parser.add_argument('--hidden_dim', type=int, default=768)
parser.add_argument('--clip_vision_width', type=int, default=1024)
parser.add_argument('--frequency_encoder_layer', type=int, default=2)
parser.add_argument('--decoder_layer', type=int, default=4)
parser.add_argument('--num_heads', type=int, default=12)


#And there is no use of dwt in this CLIPMODEL class
    def forward_FFN(self, tgt):
        tgt2 = self.linear2(self.activation(self.linear1(tgt)))
        tgt = tgt + tgt2
        tgt = self.norm2(tgt)
        return tgt

**Comment by githuboflk** [comment]
> just follow the main.py

main.py contains only the evaluation code. Can you provide your training code? Or did the author share his training code with you? Others responded that the author did not seem to share the training code with them either.

**Comment by githuboflk** [comment]
@Cwyxx

**Comment by Cwyxx** [comment]
@githuboflk sorry, I don't train the model. I just run it and find that it also can't generalize to unseen generative model like FLUX, HYDiT and so on.

**Comment by Cwyxx** [comment]
> [@Cwyxx](https://github.com/Cwyxx) Thanks, Forgive me for being a newbie, I didn't see the code referenced by dwt, please give me some guidance again. #I see that this paragraph is not used main.py frequency parser.add_argument('--hidden_dim', type=int, default=768) parser.add_argument('--clip_vision_width', type=int, default=1024) parser.add_argument('--frequency_encoder_layer', type=int, default=2) parser.add_argument('--decoder_layer', type=int, default=4) parser.add_argument('--num_heads', type=int, default=12)
> 
> #And there is no use of dwt in this CLIPMODEL class def forward_FFN(self, tgt): tgt2 = self.linear2(self.activation(self.linear1(tgt))) tgt = tgt + tgt2 tgt = self.norm2(tgt) return tgt

其实我也不知道这是干什么的，我的任务是跑通这个模型即可，抱歉没有帮到你
